invokeai/app/services/invocation_cache/invocation_cache_memory.py

```diff
@@ -32,80 +42,89 @@ def start(self, invoker: Invoker) -> None:
-            return item[0]
-        self._misses += 1
+        with self._lock:
+            if self._max_cache_size == 0 or self._disabled:
+                return None
+            item = self._cache.get(key, None)
+            if item is not None:
+                self._hits += 1
+                self._cache.move_to_end(key)
+                return item.invocation_output
+            self._misses += 1
+            return None
 
     def save(self, key: Union[int, str], invocation_output: BaseInvocationOutput) -> None:
-        if self._max_cache_size == 0 or self._disabled:
+        with self._lock:
+            if self._max_cache_size == 0 or self._disabled or key in self._cache:
+                return
+            # If the cache is full, we need to remove the least used
+            number_to_delete = len(self._cache) + 1 - self._max_cache_size
+            self._delete_oldest_access(number_to_delete)
+            self._cache[key] = CachedItem(time(), invocation_output, invocation_output.json())
+
+    def _delete_oldest_access(self, number_to_delete: int) -> None:
+        number_to_delete = min(number_to_delete, len(self._cache))
+        if self._max_cache_size == 0 or number_to_delete < 1:
             return
+        for _ in range(number_to_delete):
+            self._cache.popitem(last=False)
 
-        if key not in self._cache:
-            self._cache[key] = (invocation_output, invocation_output.json())
-            self._cache_ids.put(key)
-            if self._cache_ids.qsize() > self._max_cache_size:
-                try:
-                    self._cache.pop(self._cache_ids.get())
-                except KeyError:
-                    # this means the cache_ids are somehow out of sync w/ the cache
-                    pass
-
-    def delete(self, key: Union[int, str]) -> None:
+    def _delete(self, key: Union[int, str]) -> None:
         if self._max_cache_size == 0:
             return
-
         if key in self._cache:
             del self._cache[key]
 
-    def clear(self, *args, **kwargs) -> None:
-        if self._max_cache_size == 0:
-            return
-
-        self._cache.clear()
-        self._cache_ids = Queue()
-        self._misses = 0
-        self._hits = 0
+    def delete(self, key: Union[int, str]) -> None:
+        with self._lock:
+            return self._delete(key)
 
-    def create_key(self, invocation: BaseInvocation) -> int:
+    def clear(self, *args, **kwargs) -> None:
+        with self._lock:
+            if self._max_cache_size == 0:
+                return
+            self._cache.clear()
+            self._misses = 0
+            self._hits = 0
+
+    @staticmethod
+    def create_key(invocation: BaseInvocation) -> int:
         return hash(invocation.json(exclude={"id"}))
 
     def disable(self) -> None:
-        if self._max_cache_size == 0:
-            return
-        self._disabled = True
+        with self._lock:
+            if self._max_cache_size == 0:
+                return
```

Comment:
Not for this PR, but I'm noticing some weird interactions between `_max_cache_size` and `_disabled` in this class. @psychedelicious, maybe you can shed some light on the intended behaviour?

Things I find weird:
1. Some methods are skipped if `self._max_cache_size == 0 or self._disabled`. Others are only skipped if `self._max_cache_size == 0` (e.g. delete(), _delete_by_match())
2. If `self._max_cache_size == 0` you can't call `enable()`/`disable()`.

Can we settle on a single definition for enabled/disabled? Perhaps:
```
def _is_enabled(self):
    return not self._disabled and self._max_cache_size > 0
```

Reply:
Max cache size is used to disable the entire cache code path. It's a config setting.  Requires application restart to change. 

Disable is a runtime flag the user has control over (enable/disable button in the ui)

When the feature is config-disabled, we can skip all cache logic. That's why the check for max size is there.

When the cache is user-disabled, we still need to allow deletes and clears to occur, but should otherwise interact with the cache. That's why the check for the disabled flag is present in only some of the functions. 

I'm happy for this to be changed. This feature is currently in RC so if we need to make changes now is the time to do it.

Reply:
I think if we pasted this explanation in the class docstring, that would clear up the confusion. @lillekemiker , feel free to add this in your PR if you want, or I can create a follow-up.

Reply:
Let's put this in a followup